src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java

```diff
@@ -551,69 +527,78 @@ private void setCachedResultAndResumeToken() {
+   *       POLL_MAX_BATCH_SIZE_CONFIG}.
    *   <li>If copying data and all data has been copied and there is a cached result return the
    *       cached result.
-   *   <li>Otherwise, return the next result from the change stream cursor. Creating a new cursor if
-   *       necessary.
+   *   <li>Otherwise, return the next batch of results from the change stream cursor. Creating a new
+   *       cursor if necessary.
    * </ol>
    *
    * @return the next document
    */
-  private Optional<BsonDocument> getNextDocument() {
+  private List<BsonDocument> getNextBatch() {
+    List<BsonDocument> batch = new ArrayList<>();
+
+    long maxBatchSize = sourceConfig.getInt(POLL_MAX_BATCH_SIZE_CONFIG);
     if (isCopying) {
       assertNotNull(copyDataManager);
-      Optional<BsonDocument> result = copyDataManager.poll();
-      if (result.isPresent() || copyDataManager.isCopying()) {
-        return result;
+      if (copyDataManager.isCopying()) {
+        Optional<BsonDocument> result;
+        do {
+          result = copyDataManager.poll();
+          result.ifPresent(batch::add);
+        } while (result.isPresent() && batch.size() < maxBatchSize);
+        return batch;
       }
 
+      // Copying finished - mark copying ended and add cached result
       isCopying = false;
       LOGGER.info("Finished copying existing data from the collection(s).");
       if (cachedResult != null) {
-        result = Optional.of(cachedResult);
+        batch.add(cachedResult);
         cachedResult = null;
-        return result;
       }
     }
 
     if (cursor == null) {
       initializeCursorAndHeartbeatManager();
+    } else if (cursor.getServerCursor() == null) {
+      LOGGER.info("Cursor has been closed by the server - reinitializing");
+      invalidateCursorAndReinitialize();
     }
 
-    if (cursor != null) {
-      try {
-        BsonDocument next = cursor.tryNext();
-        // The cursor has been closed by the server
-        if (next == null && cursor.getServerCursor() == null) {
-          invalidateCursorAndReinitialize();
-          next = cursor != null ? cursor.tryNext() : null;
-        }
-        return Optional.ofNullable(next);
-      } catch (MongoException e) {
-        closeCursor();
-        if (isRunning) {
-          if (sourceConfig.tolerateErrors() && changeStreamNotValid(e)) {
-            cursor = tryRecreateCursor(e);
-          } else {
-            LOGGER.info(
-                "An exception occurred when trying to get the next item from the Change Stream", e);
-          }
+    try {
+      BsonDocument next;
+      do {
+        next = cursor.tryNext();
+        if (next != null) {
+          batch.add(next);
         }
-        return Optional.empty();
-      } catch (Exception e) {
-        closeCursor();
-        if (isRunning) {
-          throw new ConnectException("Unexpected error: " + e.getMessage(), e);
+      } while (next != null && batch.size() < maxBatchSize && cursor.available() > 0);
```

Comment:
Do while loops simplify the logic - so get all the data until there is no data or we reach maxBatchSize